Aqara Button DTH - add support for model WXKG12LM

veeceeoh opened the pull request:
Aqara Button DTH code reworked to add support for the new Aqara Button model WXKG12LM.

Supports all four button event messages, click, double-click, hold, and shake, emulating a 3-button device as follows:

Button action | Event sent to hub
-----|-----
single-click | `button 1 pushed`
hold | `button 1 held`
double-click | `button 2 pushed`
shake | `button 3 pushed`

On any of the above button events, `lastPressed` and `lastPressedCoRE` events with a date/time stamp are also sent, the former for main tile display, and the latter being useful for webCoRE users.

Note that although the button sends a message on release after being held, SmartThings no longer supports a `button released` event, so only custom `lastReleased` and `lastReleasedCoRE` events are generated, to be used by webCoRE users or people using custom Smart Apps that can access non-standard attribute events.

Other changes:
• removed `button released` event from previous DTH code
• added Advanced Settings section in preference page
• added info log and debug log message display toggle settings to Advanced Settings section, with default of info log display turned on and debug log display off
• grouped log messages into info and debug categories and changed to use new log message display calls `displayDebugLog()` and `displayInfoLog()`
• edited some log message text for clarity
• removed redundant and non-useful log messages
• added some new log messages related to the code added for the new Aqara model 
• changed `lastCheckinDate` event attribute to `lastCheckinCoRE`
• switched to use more accurate _Epoch Time Stamp_ method of now() for all ____CoRE events
• switched method for human readable date stamp of events to directly call `formatDate()` routine in the DTH

veeceeoh commented:
Please note that this code NOT yet been tested. I will have a chance to do that later today - as long as the buttons delivered to me were model WXKG12LM.

veeceeoh commented:
Okay, I needed to fix some syntax errors, and did a bunch of testing and it's working now!

![screen shot 2018-03-23 at 9 13 27 pm](https://user-images.githubusercontent.com/33558908/37860264-1410825c-2edf-11e8-80d2-206cfe73cf52.jpg)

**Some initial impressions**:
* Interestingly, I couldn't get it to pair with my Hubitat hub. After 9-10 attempts, I gave up. I haven't opened tried the second new Aqara button I got, so I will try again soon
* I've only just paired one button once with my SmartThings hub. It didn't take very long, but it was recognized as a "Thing". I used the pairing in/outCluster info to edit the fingerprint in the DTH, so I'm going to try pairing again to see if it can be recognized. If I can get one of them paired with my Hubitat hub, I can get the whole correct in/outCluster information for the fingerprint.
* The button hold feature is hardware-based, so the time is not editable. It seems you need to hold it for about 2.5-3 seconds before the held message is sent.
* I sometimes don't see the release message happen after the hold message. So I am not sure it can be relied on for webCoRE Piston use.
* I also sometimes didn't see the hold message sent, even when holding for 5+ seconds. Not sure why.
* Single-clicks and double-clicks that are _too_ fast aren't recognized (no message sent). This just means it will take some personal "training" to get used to the right length of press for a single click, and the speed of a double-click. Again, these timings cannot be adjusted. This button just works the way it was designed, and that is what we have to work with.

veeceeoh commented:
Also, upon my first pairing of the button, I noticed this:

![screen shot 2018-03-23 at 8 46 52 pm](https://user-images.githubusercontent.com/33558908/37860322-d4e3581a-2edf-11e8-80f9-b6aec7581987.jpg)

Not sure if it's a display issue in IDE or just not all of the model name was captured during pairing.

veeceeoh commented:
Whoo hoo! I added another fingerprint with `model: "lumi.sensor_swit"` and it actually paired correctly!

However, something is causing an error during the install() / configure() steps. I'm nearly there...

veeceeoh commented:
Figured out reason for the error during pairing, fixed it, and learned a few good things.

This code is pretty much ready to use, but needs testing by others who have the model WXKG12LM and also the original WXKG11LM model to make sure it works too. (I can only test the WXKG11LM functionality with one of my wall-mounted smart switches, which did work with the previous Aqara button DTH.)

veeceeoh commented:
One last update - I did some testing with my 2-button wall-surface-mounted Aqara Smart Light Switch (model WXKG02LM) because it works the same as the original Aqara square button (model WXKG11LM) and discovered I had forgot to update the name of a function call. So that's fixed.

Then I added fingerprint info for both of the Aqara Smart Light Switch models (WXKG01LM and WXKG02LM). Note that for the 2-button model, SmartThings treats it as a SINGLE button, because the Zigbee API strips out the `endpoint` number from the button press message, so all three types of button presses (left, right, both) are received by the DTH as `on/off: 1` which is VERY dissapointing (and one of the reasons I got a Hubitat hub, because it doesn't ignore any type of Zigbee messages).

I also fixed an error that could happen if someone pairs a button using the manual "catchall" method (which I had to use for my Aqara Smart Light Switch).

veeceeoh commented:
I just tried pairing the new Aqara Button I've been testing in SmartThings over on my Hubitat hub, and it only took 3 tries.

This makes me wonder if at the factory Xiaomi / Aqara devices are set on a specific Zigbee channel, and they may have trouble "finding" hubs on other certain Zigbee channels, until the Xioami devices have changed their channel to something closer... ???

I am going to try to find out what the default Zigbee channel is for both the rev 1 and rev 2 Xiaomi Gateways.

veeceeoh commented:
@xtianpaiva - have you had a chance to try this updated DTH?

[Direct link to the new code here](https://raw.githubusercontent.com/bspranger/Xiaomi/updated-aqara-dth/devicetypes/bspranger/xiaomi-aqara-button.src/xiaomi-aqara-button.groovy).

@bspranger - Since I removed the button "release" event (because it's no longer supported in SmartThings,) which means the toggle mode and wait until release timer are also gone, I just want to make sure you're okay with this new DTH version!

xtianpaiva commented:
sorry @veeceeoh i was on vacation last week, was testing the DTH today and it looks good to me!

veeceeoh commented:
New update!

Using custom a custom attribute, the main tile display now shows the state of the button, with a new custom button icon!

On any button action, appropriate text is displayed with a blue background and custom pushed button icon for 1 second, and is then reset back to grey with "released" label and released button icon. For model WXKG12LM, when the button is held, the main tile will remain displaying "held" with blue background until the button is physically released. Screenshots below...

**Changes**
• added custom tileAttribute `buttonStatus` to main tile to handle display of button actions, with grey color "Released" state, and all other states (pushed, held, single-clicked, double-clicked, & shaken) all displayed with corresponding label in standard SmartThings

 blue color.
• created and added new custom icons for "released" and all "pushed" states
• refactored virtual button `push()` function code

![img_2370](https://user-images.githubusercontent.com/33558908/38392850-c769925a-38dd-11e8-8320-b0b85281bfc5.PNG)

![img_2365](https://user-images.githubusercontent.com/33558908/38392855-d21eba40-38dd-11e8-9adb-74678dd8c48a.PNG)
